Add some elaboration in TiKV doc (#40)
tikv/sig-transaction: doc/tikv/README.md

## Optimistic and pessimistic transactions

> TiKV supports two transaction models. Optimistic transactions were implemented first and often when TiKV folks don't specify optimistic or pessimistic, they mean optimistic by default. In the optimistic model, reads and writes are built up locally. All writes are sent together in a prewrite. During prewrite, all keys to be written are locked. If any keys are locked by another transaction, return to the client. If all prewrites succeed, the client sends a commit message.

Reviewer: 'the' is not needed here, if we do use it then it should be 'return them to the client'

> Pessimistic transactions are the default in TiKV since 3.0.8. In the pessimistic model, there are *locking reads* (from `SELECT ... FOR UPDATE` statements), these read a value and lock the key. This means that reads can block. SQL statements which cause writes, lock the keys as they are executed. Writing to the keys is still postponed until the prewrite. Prewrite and commit works

Reviewer: 'the' is incorrect here. I can't even explain why, English just sucks.

TODO for each: why? implications, benefits

### Timestamps are supplied by the client

This decision benefits "user experience", performance and simplicity.

First, it gives users more control over the order of concurrent transactions.

For example, a client commits two transactions: T1 and then T2. 
If timestamps are supplied by the user, it can assure that T1 won't read any effects of T2 if T1's timestamp is smaller than T2's.
While if we let TiKV get the timestamp, the user cannot get this guarantee because the order of processing T1 and T2 is nondeterministic.

Second, it simplifies the system. Otherwise we have to let TiKV maintain states of all active transactions. 

> Third, it is beneficial for performance. Letting TiKV maintain states of active transactions would lead to extra network communication. Large volumn of transactions could overburden TiKV server. In addition, GC of inactive transactions is a problem.

Reviewer: ```suggestion
Third, it is beneficial for performance. Letting TiKV maintain states of active transactions would lead to extra network communication. Large volume of transactions could overburden TiKV server. In addition, GC of inactive transactions is a problem.
```

"Letting TiKV maintain states of active transactions would lead to extra network communication" I don't think this is true - it should require at most the same amount of network communication, but would allow some optimisations which would reduce network overheads.

TODO: further elaboration

### All timestamps are unique

> This no longer holds. 1PC and Async commit could break this guarantee.

Reviewer: To clarify, I think this does still hold unless the user opts-in to async commit.

> Multiple transactions may have identical commit timestamps. However one transaction must have distinct start_ts and commit_ts.

Reviewer: start_ts of all transactions must still be unique too

### Reads never fail

> Reads never fail in the Read Committed level. It will always read the latest committed version.

Reviewer: ```suggestion
Reads never fail in the read committed level. The client will always read the most recent committed version.
```

> Read requests can return Err in the Snapshot Isolation level if the key is locked with `lock_ts` < `read_ts`. The the client will try to resolve the lock and retry until it succeeds.

Reviewer: I would clarify:

```suggestion
Read requests can return `KeyError` in the snapshot isolation level if the key is locked with `lock_ts` < `read_ts`. Then the client can try to resolve the lock and retry until it succeeds.
```

> ### TiKV nodes do not communicate with each other, only with a client

Reviewer: To clarify, they don't communicate at the transaction layer. Nodes withing the same Raft group do communicate to ensure consensus at the Raft level

### The transaction layer does not know about region topology, in particular, it does not treat regions on the same node differently to other regions

A TiKV instance does not have to know the topology. The client makes sure any request is sent to the right TiKV node that owns the data involved in the request.

The design decouples transaction logic and physical data distribution. It makes shceduling more flexible and elastic. 
Imagine a redistribution of regions among a TiKV cluster that does not require any downtime or maintainance to either clients or TiKV instances.
PD as the scheduler can ask TiKV to redistribute regions, and send the latest region info to clients.

The overhead caused by such decoupling is extra network communication. Though clients must acquire regions' and TiKV stores' addresses from PD, these information be cached locally. If topology changes, client may failed some request and retry to refresh its cache. A long-live client should suffer little from it.

### If committing the primary key succeeds, then committing the secondary keys will never fail.

> Even if it fails, the lock of a secondary key contains information of its primary key. Any transactions that meets the lock can recognize its state by reading the primary key and help commit the secondary key.

Reviewer: ```suggestion
Even if the commit message sent to the secondary key fails, the lock of a secondary key contains information of its primary key. Any transactions that meets the lock can recognize its state by reading the primary key and help commit the secondary key.
```

TODO for each: why? implications, benefits

### Timestamps are supplied by the client

This decision benefits "user experience", performance and simplicity.

First, it gives users more control over the order of concurrent transactions.

For example, a client commits two transactions: T1 and then T2. 
If timestamps are supplied by the user, it can assure that T1 won't read any effects of T2 if T1's timestamp is smaller than T2's.
While if we let TiKV get the timestamp, the user cannot get this guarantee because the order of processing T1 and T2 is nondeterministic.

Second, it simplifies the system. Otherwise we have to let TiKV maintain states of all active transactions. 

Third, it is beneficial for performance. Large volume of transactions could overburden TiKV server. In addition, GC of inactive transactions is a problem.

TODO: further elaboration

### All timestamps are unique

TODO

It's true in previous versions of TiKV. Enabling 1PC or Async commit features could break this guarantee.

> Multiple transactions may have identical commit timestamps. Start timestamps are still unique. One transaction must have distinct start_ts and commit_ts.

Reviewer 2: It's mostly correct. However, we don't define a transaction precisely. So if we take failed transactions into consideration, the start_ts and commit_ts of a failed transaction are the same 😂 (the commit_ts of a rollback record equals to its start_ts)

> ### Reads never fail

Reviewer 2: This title may be misunderstanding to readers?

Reads never fail in the read committed level. The client will always read the most recent committed version.

Read requests can return `KeyError` in the snapshot isolation level if the key is locked with `lock_ts` < `read_ts`. Then the client can try to resolve the lock and retry until it succeeds.

### TiKV nodes do not communicate with each other, only with a client

> TiKV instances do not have to know each other at the transaction layer. In the raft-store layer, they communicate with peers in the same Raft group.

Reviewer 2: This is also not true in the transaction layer. TiKV nodes need to send "wait-for" edges to the deadlock detector node.
